File: Support/bin/fcshd.rb
#!/usr/bin/env ruby -wKU
# encoding: utf-8

module FCSHD

BUN_SUP = ENV['TM_BUNDLE_SUPPORT']

require 'xmlrpc/client'
#require 'Logger'
require ENV['TM_SUPPORT_PATH'] + '/lib/escape'
require ENV['TM_SUPPORT_PATH'] + '/lib/web_preview'

require ENV['TM_BUNDLE_SUPPORT'] + '/bin/fcshd_server'

# @logger = Logger.new('/tmp/fcshd/gui.log')
# @logger.level = Logger::DEBUG

def self.status
	if FCSHD_SERVER.running
		puts "Running"
	else
		puts "Stopped"
	end
end

def self.generate_view
		
		puts html_head(:window_title => "ActionScript 3", :page_title => "fcshd", :sub_title => "__" );

		puts	"<link rel='stylesheet' href='file://#{e_url(BUN_SUP)}/css/fcshd.css' type='text/css' charset='utf-8' media='screen'>"
		puts  "<script src='file://#{e_url(BUN_SUP)}/js/fcshd.js' type='text/javascript' charset='utf-8'></script>"
		puts "<div id='script-path'>#{BUN_SUP}/bin/</div>"
		puts "
		<h2><div id='status'>Checking daemon status</div></h2>
		<div id='controls'>
		  <a id='refresh' href='javascript:refreshStatus()' title='Check daemon status'>Check Status</a><br/>
		  <a id='toggle' href='javascript:toggleClick();'>Toggle State</a><br/>
		</div>"
		
		compiler_state = FCSHD_SERVER.running ? "up" : "down"
		puts '<script type="text/javascript" charset="utf-8">setState("'+compiler_state+'")</script>'
end

def self.stop_server
	return unless FCSHD_SERVER.running
	FCSHD_SERVER.stop_server
	sleep 0.5
	status
end

def self.start_server
	return if FCSHD_SERVER.running
	FCSHD_SERVER.start_server
	
	#Give up from waiting if it's taking too long
	start_time = Time.now.to_i
	while !FCSHD_SERVER.running
	 sleep 0.5
	 break if Time.now.to_i - start_time > 1000 * 10
	end

end

def self.success
    print "<script type='text/javascript' charset='utf-8'>
      if( document.getElementById('status').className != 'fail'){
        document.getElementById('status').className='success'
        document.getElementById('status').innerHTML='Success'
      }
    </script>"  
end

def self.fail
  print "<script type='text/javascript' charset='utf-8'>
      document.getElementById('status').innerHTML='Compilation Failed'
      document.getElementById('status').className='fail'
  </script>"
end

def self.close_window
  print "<script type='text/javascript' charset='utf-8'>
        window.close();
  </script>"
end

end

def run
  if ARGV[0] == "-success" 
    FCSHD.success
  elsif ARGV[0] == "-fail"
      FCSHD.fail
    elsif ARGV[0] == "-status"
      FCSHD.status
    elsif ARGV[0] == "-start"
      FCSHD.start_server
    elsif ARGV[0] == "-stop"
      FCSHD.stop_server
    elsif ARGV[0] == "-view"
      FCSHD.generate_view
  else
    "No command given"
  end
end
run

File: Support/js/fcshd.js
document.title = "Flex shell daemon status"

function setState(newState)
{
    console.log('state -> ' + newState)

    var statusMsg = ""
    var linkMsg = ""
    switch (newState)
    {
        case "up":
        	statusMsg = "Daemon is running"
        	linkMsg = "Stop"
        break

        case "down":
        	statusMsg = "Daemon is down"
        	linkMsg = "Start"
        break

        case "launching":
        	statusMsg = "Launching..."
        break

		case "stopping":
			statusMsg = "Daemon is stopping..."
		break;

		case "starting":
			statusMsg = "Daemon is starting..."
		break;
		
        case "unknown":
        	statusMsg = "Unknown..."
        break;

    }

    document.getElementById('status').innerHTML = statusMsg;
    document.getElementById('status').className = newState;
    document.getElementById('toggle').innerHTML = linkMsg;
    document.getElementById('toggle').className = newState;

}

function refreshStatus()
{
    var scriptElt = document.getElementById('script-path')
    var scriptPath = scriptElt.firstChild.nodeValue
	console.log( scriptPath + "fcshd.rb");
    TextMate.system('echo `"' + scriptPath + 'fcshd.rb" -status`', 
	    function(e) {
			console.log(e.outputString);
	        if (e.outputString.match("Running"))
			{
	            setState('up')

	        }
			else if (e.outputString.match("Stopped"))
			{
	            setState('down')
	        } 
			else
			{
				setState('unknown');
			}
	    })
}

function toggleClick()
{
    var scriptElt = document.getElementById('script-path')
    var scriptPath = scriptElt.firstChild.nodeValue

    var linkElt = document.getElementById('toggle')
    switch (linkElt.className)
	{
        case 'up':

			setState('stopping');
        	console.log('stopping');

			TextMate.system('"' + scriptPath + 'fcshd.rb" -stop', 
        		function(e) {
        		    console.log('stdout: ' + e.outputString)
        		    console.log('stderr: ' + e.errorString)
        		    if (e.outputString.match("Stopped"))
					{
        		        setState('down')
        		    } 
					else
					{
        		        console.log('but no match: "' + e.outputString + '" vs "STOPPED"')
        		    }
        		});
        break;

        case 'down':

			console.log('launching')
			setState('launching')
	
			console.log('"' + scriptPath + 'fcshd.rb" -start');
			TextMate.system('"' + scriptPath + 'fcshd.rb" -start', 
				function(e) {

					console.log('stdout: ' + e.outputString)
					console.log('stderr: ' + e.errorString)
	
					if (e.outputString.match("Running"))
					{
						setState('up')
					} 
					else if (e.outputString.match("Stopped"))
					{
						setState('down')
					}
					else
					{
						setState('unknown')
					}
				})
	
		break;
 
		case 'launching':
			console.log('launching')
		break;

		case 'unknown':
			console.log('unknown')
		break;
		
	}

}